tensorcircuit/applications/optimization.py

```diff
@@ -0,0 +1,328 @@
+"""
+modules for QUBO problems in QAOA
+"""
+
+from typing import List, Callable, Any
+
+import tensorcircuit as tc
+import numpy as np
+from functools import partial
+import tensorflow as tf
+import scipy.optimize as optimize
+
+from ..cons import backend
+from ..templates.blocks import QAOA_ansatz_for_Ising
+from ..templates.conversions import QUBO_to_Ising
+from tensorflow.python.ops.numpy_ops import np_config
+
+np_config.enable_numpy_behavior()
+
+Circuit = Any
+Tensor = Any
+
+
+def Ising_loss(c: Circuit, pauli_terms: List[list], weights: list) -> float:
+    """
+    computes the loss function for the Ising model based on a given quantum circuit,
+    a list of Pauli terms, and corresponding weights.
+    The offset is ignored.
+
+    :param c: A quantum circuit object generating the state.
+    :param pauli_terms: A list of Pauli terms, where each term is represented as a list of 0/1 series.
+    :param weights: A list of weights corresponding to each Pauli term.
+    :return loss: A real number representing the computed loss value.
+    """
+    loss = 0.0
+    for k in range(len(pauli_terms)):
+        term = pauli_terms[k]
+        index_of_ones = []
+
+        for l in range(len(term)):
+            if term[l] == 1:
+                index_of_ones.append(l)
+
+        # Compute expectation value based on the number of qubits involved in the Pauli term
+        if len(index_of_ones) == 1:
+            delta_loss = weights[k] * c.expectation_ps(z=[index_of_ones[0]])
+            # Compute expectation value for a single-qubit Pauli term
+        else:
+            delta_loss = weights[k] * c.expectation_ps(
+                z=[index_of_ones[0], index_of_ones[1]]
+            )
+            # Compute expectation value for a two-qubit Pauli term
+
+        loss += delta_loss
+
+    return backend.real(loss)
+
+
+def QAOA_loss(
+    nlayers: int, pauli_terms: List[list], weights: list, params: list, mixer: str = "X"
+) -> float:
+    """
+    computes the loss function for the Quantum Approximate Optimization Algorithm (QAOA) applied to the Ising model.
+
+    :param nlayers: The number of layers in the QAOA ansatz.
+    :param pauli_terms: A list of Pauli terms, where each term is represented as a list of 0/1 series.
+    :param weights: A list of weights corresponding to each Pauli term.
+    :param params: A list of parameter values used in the QAOA ansatz.
+    :return: The computed loss value.
+    """
+    c = QAOA_ansatz_for_Ising(params, nlayers, pauli_terms, weights, mixer=mixer)
+    # Obtain the quantum circuit using QAOA_from_Ising function
+
+    return Ising_loss(c, pauli_terms, weights)
+    # Compute the Ising loss using Ising_loss function on the obtained circuit
+
+
+def QUBO_QAOA(
```

Comment:
this looks like a more comprehensive version of the same function that appears in QAOA_funcs.py